### PR TITLE
DAOS-8048 placement: Not put same grp shards in the same domain

### PR DESCRIPTION
**Pull request by wangdi1**
Add doms_cur_grp_used bits array in get_targets to make sure
do not put shards within the same group in the same domain.

Signed-off-by: Di Wang <di.wang@intel.com>

### CODE REVIEW

File: src/placement/jump_map.c [hunk 1]
```diff
@@ -381,13 +386,38 @@ get_target(struct pool_domain *curr_dom, struct pool_target **target,
...
 				/* Skip the domain whose targets are used up */
 				for (idx = start_dom; idx <= end_dom; ++idx) {
-					if (isclr(dom_occupied, idx))
+					/* Only reused the domain, if there are still targets
+					 * avaible (not being used) within this domain, and the
+					 * domain has not being used by current group yet. so
+					 * 1. there won't be multiple shards in the same target.
+					 * 2. there won't be multiple shards within same group
+					 *    are in the same domain.
+					 */
```
Comment (NiuYawei) [inline src/placement/jump_map.c:418]: What about SX object? And the isset_range() calls makes me worrying the performance.

File: src/placement/jump_map.c [hunk 2]
```diff
@@ -705,8 +792,16 @@ get_object_layout(struct pl_jump_map *jmap, struct pl_obj_layout *layout,
...
-				rc = remap_alloc_one(remap_list, k, target,
-						     false);
+
+				if (remap_grp_used == NULL) {
+					remap_grp_used = remap_gpu_alloc_one(&dgu_remap_list,
+									     dom_cur_grp_used);
+					if (remap_grp_used == NULL)
+						D_GOTO(out, rc = -DER_NOMEM);
+					realloc_grp_used = true;
+				}
```
Comment (NiuYawei) [inline src/placement/jump_map.c:802]: It's really hard to follow this, why don't you just simply allocate per-RDG "used domain bitmap" for all groups at the very beginning (Just like what we did for other bitmaps)?

Comment (wangdi1) [reply src/placement/jump_map.c:802]: Hmm, I was trying to only allocate once for each layout generation normal, i.e. only re-allocate the grp tracking bitmap for remap case. Since this is on the hot path, I have to be cautious to avoid extra allocation for multiple group case.

File: src/placement/jump_map.c (hunk 1 above)
Comment (wangdi1) [reply src/placement/jump_map.c:418]: SX will only have 1 shard per group, what do you mean here?  

Actually the comments is a bit misleading, it actually means it will try not to put multiple shards within same group to be put into the same domain. But we do not promise that, if the group has been used all domain during rebuild, then we might still put multiple shards in the same domain.

File: src/placement/jump_map.c (hunk 2 above)
Comment (NiuYawei) [reply src/placement/jump_map.c:828]: yes, I see your point now. Growing the 'uesd domain bitmap' on-demand could be better.

**Review by NiuYawei — DISMISSED** [review]

File: src/placement/jump_map.c (hunk 1 above)
Comment (wangdi1) [reply src/placement/jump_map.c:436]: hmm, it should add much impacts for GX layout generation, might be some extra jump for some extrem corner case, but that is unavoided for GX anyway.

**Review by ryon-jensen — DISMISSED** [review]

**Review by liuxuezhao — APPROVED** [review]

**Review by ryon-jensen — APPROVED** [review]
